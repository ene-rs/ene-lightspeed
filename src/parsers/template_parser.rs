use anyhow::bail;
use pest::{iterators::Pair, Parser};
use pest_derive::Parser;

#[derive(Parser)]
#[grammar = "./parsers/template_grammar.pest"]
pub struct TemplateParser;

impl TemplateParser {
    pub fn parse_template(template: &str) -> anyhow::Result<TemplateEntity> {
        let pairs = TemplateParser::parse(Rule::entity, template)?
            .next()
            .unwrap();
        let entity_content = pairs
            .into_inner()
            .map(TemplateParser::inner_parse_rule)
            .collect::<anyhow::Result<Vec<TemplateEntityContent>>>()?;
        Ok(TemplateEntity { entity_content })
    }

    fn inner_parse_rule(pair: Pair<Rule>) -> anyhow::Result<TemplateEntityContent> {
        match pair.as_rule() {
            Rule::content => Ok(TemplateEntityContent::Static {
                content: pair.as_str().to_string(),
                start: pair.as_span().start(),
                end: pair.as_span().end(),
            }),
            Rule::attribute => Ok(TemplateEntityContent::Attribute {
                content: pair
                    .clone()
                    .into_inner()
                    .next()
                    .unwrap()
                    .as_str()
                    .to_string(),
                start: pair.as_span().start(),
                end: pair.as_span().end(),
            }),
            Rule::filter => Ok(TemplateEntityContent::Filter {
                content: pair
                    .clone()
                    .into_inner()
                    .map(TemplateParser::inner_parse_filter)
                    .collect::<anyhow::Result<Vec<FilterContent>>>()?,
            }),
            Rule::entity_content => {
                TemplateParser::inner_parse_rule(pair.into_inner().next().unwrap())
            }
            s => bail!("Unexpected rule: {:?}", s),
        }
    }

    fn inner_parse_filter(pair: Pair<Rule>) -> anyhow::Result<FilterContent> {
        match pair.as_rule() {
            Rule::content => Ok(FilterContent::Static {
                content: pair.as_str().to_string(),
                start: pair.as_span().start(),
                end: pair.as_span().end(),
            }),
            Rule::filter_attribute => Ok(FilterContent::FilterAttribute {
                content: pair
                    .clone()
                    .into_inner()
                    .next()
                    .unwrap()
                    .as_str()
                    .to_string(),
                start: pair.as_span().start(),
                end: pair.as_span().end(),
            }),
            Rule::filter_content => {
                TemplateParser::inner_parse_filter(pair.into_inner().next().unwrap())
            }
            s => bail!("Unexpected rule: {:?}", s),
        }
    }
}

#[derive(Debug)]
pub struct TemplateEntity {
    pub entity_content: Vec<TemplateEntityContent>,
}

#[derive(Debug)]
pub enum TemplateEntityContent {
    Static {
        content: String,
        start: usize,
        end: usize,
    },
    Attribute {
        content: String,
        start: usize,
        end: usize,
    },
    Filter {
        content: Vec<FilterContent>,
    },
}

#[derive(Debug)]
pub enum FilterContent {
    Static {
        content: String,
        start: usize,
        end: usize,
    },
    FilterAttribute {
        content: String,
        start: usize,
        end: usize,
    },
}

#[cfg(test)]
mod tests {
    use crate::parsers::template_parser::TemplateParser;

    #[test]
    fn test_parser() {
        let string_test = std::fs::read_to_string("/Users/abdullahsabaaallil/Desktop/ene-lightspeed/template-rs/src/models/entity_name.rs").unwrap();
        let template_entity = TemplateParser::parse_template(&string_test);
        insta::assert_debug_snapshot!(template_entity);
    }
}
