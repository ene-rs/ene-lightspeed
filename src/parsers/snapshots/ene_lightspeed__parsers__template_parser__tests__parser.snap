---
source: src/parsers/template_parser.rs
expression: template_entity
---
Ok(
    TemplateEntity {
        entity_content: [
            Static {
                content: "\nuse serde::Serialize;\nuse serde::Deserialize;\nuse uuid::Uuid;\nuse crate::error::Error;\nuse crate::models::PaginatedResult;\nuse serde::Serializer;\nuse crate::controllers::entity_name_controller::{AddEntityNamePayload, UpdateEntityNamePayload};\n\n\n#[derive(Debug, Clone, Serialize, Deserialize)]\n#[serde(rename_all = \"camelCase\")]\npub struct EntityName {",
                start: 18,
                end: 370,
            },
            Attribute {
                content: "\n    pub attribute_name: AttributeType,\n    pub attribute2_name: Attribute2Type,",
                start: 370,
                end: 511,
            },
            Static {
                content: "\n}\n\nimpl EntityName {\n    pub fn new(payload: AddEntityNamePayload) -> Result<Self, Error> {\n        Ok(Self {\n            primary_key: payload.primary_key,",
                start: 511,
                end: 667,
            },
            Attribute {
                content: "\n            attribute_name: payload.attribute_name,\n            attribute2_name: payload.attribute2_name,",
                start: 667,
                end: 845,
            },
            Static {
                content: "\n        })\n    }\n\n    pub fn update(payload: UpdateEntityNamePayload) -> Result<Self, Error> {\n        Ok(Self {\n            primary_key: payload.primary_key,",
                start: 845,
                end: 1004,
            },
            Attribute {
                content: "\n            attribute_name: payload.attribute_name,\n            attribute2_name: payload.attribute2_name,",
                start: 1004,
                end: 1182,
            },
            Static {
                content: "\n        })\n    }\n}\n\n#[derive(Debug, Clone, Serialize, Deserialize)]\n#[serde(rename_all = \"camelCase\")]\npub struct EntityNameFilterAttributes {\n    page: Option<i32>,\n    page_size: Option<i32>,",
                start: 1182,
                end: 1376,
            },
            Filter {
                content: [
                    FilterAttribute {
                        content: "\n    filter_attribute_name: Option<FilterAttributeType>,\n    filter2_attribute_name: Option<Filter2AttributeType>,",
                        start: 1400,
                        end: 1584,
                    },
                ],
            },
            Static {
                content: "\n}\n\nimpl EntityNameFilterAttributes {",
                start: 1611,
                end: 1648,
            },
            Filter {
                content: [
                    Static {
                        content: "\n    fn is_most_specific_filter_attribute_name_filter(&self) -> bool {",
                        start: 1671,
                        end: 1741,
                    },
                    FilterAttribute {
                        content: "\n        self.filter_attribute_name.is_some() && self.filter2_attribute_name.is_some()",
                        start: 1741,
                        end: 1905,
                    },
                    Static {
                        content: "\n    }",
                        start: 1905,
                        end: 1911,
                    },
                ],
            },
            Static {
                content: "\n\n}",
                start: 1938,
                end: 1941,
            },
        ],
    },
)
